hardware-computing/README.md

```diff
@@ -1,8 +1,69 @@
+Some of these operations are critical in ensuring the security of a system.
+In this session we will explore some of those:
+ * Encryption and decryption on Intel and ARM CPUs, in particular the AES algorithm
+ * Random number generation
 
-## ARM Encryption Modules
+After that is done we have one more topic: can we trust the hardware?
+Many people posed that question, including tech giants.
+Those questions resulted in the creation of the `Trusted Computing` concept.
+We'll see what it is and how it is implemented. 
+
+## AES Instruction Sets
+
+One of the most common algorithms used for encryption and decryption is AES (Advanced Encryption Standard).
+As it's name suggests, it is the standard for securing data that cannot (should not) be intercepted over a network.
+It is used to encrypt certain files, whole disks, by password managers to store passwords, and other uses.
+Being used so much, it was only natural that the most efficient implementation was searched.
+And, generally, the most efficient implementation of an algorithm is directly on hardware.
+
+Implementation on hardware has one more benefit: it is more secure.
+If done by software, while being secure against brute-force attacks, AES can be broken using side-channel attacks.
+The hardware implementation is much harder to be affected by such attacks.
+
+Intel, AMD, ARM, Risc-V SPARC, IBM have implemented AES instruction sets in hardware.
+We will discuss the implentations from Intel and ARM.
+If interested in the implementations from other architectures, you can find them in the `Further Reading` section.
+
+If you are not familiar with how AES works, you can find details about it in the `helpers/` folder, in the `aes.md` file.
+
+### Intel AES-NI
+
+In 2008, Intel proposed an AES instruction set for the x86_64 CPUs, AES-NI (Advanced Encryption Standard New Instructions).
+AMD also implemented the instruction set for their CPUs.
+
+The instruction set provides 2 instructions for encrypting, 2 for decrypting, and 2 for key generation:
+ * AESENC performs an encryption round
+ * AESENCLAST performs the last encryption round
+ * AESDEC performs a decryption round
+ * AESDECLAST performs the last decryption round
+ * AESKEYGENASSIS aids in the key expansion phase of the algorithm
+ * AEIMC prepares a key for being used in decryption
+
+All the instructions use XMM registers, which are part of the SSE / AVX instruction sets.
+If you are unsure what those are, you can find information about them in the `helpers/` folder, in the `simd-instructions.md` file.
+
+In the case of the first 4 and the 6th instructions, 2 XMM registers are used: the first one stores the intermediary state of the cyphertext, and the second one the key used in that round.
+The 5th one also takes an immediate value, called RCON, which is used in the key generation algorithm.
```

Comment:
Is this necessary? The students will preferably get this information in the section `AES in ASM`.

Reply:
I'm not sure. I'll see when I make the challenges.